framework/include/actioncomponents/ActionComponent.h

```diff
@@ -0,0 +1,92 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+// MOOSE includes
+#include "Action.h"
+#include "ActionWarehouse.h"
+
+class PhysicsBase;
+class FEProblem;
+
+#define registerActionComponent(app_name, component_name)                                          \
+  registerMooseAction(app_name, component_name, "list_component")
+
+/**
+ * Base class for components that are defined using an action
+ */
+class ActionComponent : public Action
+{
+public:
+  static InputParameters validParams();
+
+  ActionComponent(const InputParameters & params);
+
+  virtual void act() override final;
+
+  /// Get the name of the mesh generator created by this component that generates the mesh for it
+  /// - this could be a mesh generator in the [Mesh] block
+  /// - or a mesh generator created by the component
+  const std::vector<MeshGeneratorName> & meshGeneratorNames() const { return _mg_names; }
+
+  /// Returns the subdomains for the component mesh, if any
+  const std::vector<SubdomainName> & blocks() const { return _blocks; }
```

Comment:
What are your thoughts on having something like `MeshedActionComponent` so that there aren't these interfaces for non-meshed components?

Reply:
we really need to decide on this inheritance tree :O

Reply:
This is my only suggestion to make to the tree. It can live without the change, but it's just going to be a little janky being able to call methods like `volume()` and such for 0-D components. What are your thoughts on the tree?

Reply:
I would see 0D components have a volume in the same way that 1D and 2D ones have one. They represent something 3D

Same for a "boundary", we could have a nodeset of the 1 node act as the boundary, with a "surfaceArea" for the equivalent 3D element

My thoughts is no tree until we get forced into it. Just work with helpers and interfaces, which can be virtually inherited if needed

Reply:
There'll probably be some abstract components where there is not an associated volume or such. But I'm fine with waiting - it won't be too hard to change later.